### PR TITLE
libaom: Add support for butteraugli and vmaf tunes

### PR DESCRIPTION
**Pull request by D4ndellion**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dali99** [comment]
We could also enable it by default, opinions?

**Comment by dali99** [comment]
I've enabled it by default

**Comment by dali99** [comment]
I've built it on staging as well now, seems good.

File: pkgs/development/libraries/libaom/default.nix [hunk 1]
```diff
@@ -16,6 +19,13 @@ stdenv.mkDerivation rec {
     yasm perl cmake pkg-config python3
   ];
 
+  buildInputs = lib.optionals enableButteraugli [
+    # libavif -> libaom -> libjxl -> graphviz -> gd -> libavif
+    # graphviz is only used for generating docs so it's fine to remove when building libaom
+    (libjxl.override { graphviz = null; })
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libaom/default.nix:25]: overriding packages to null is a very big anti pattern and when done wrong can cause all sorts of issues. Also this should be done in a higher level do make sure we don't mix two variants.
I would suggest to add a flag to libjxl to not build the doc instead.

Comment (dali99) [reply pkgs/development/libraries/libaom/default.nix:25]: That makes sense

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -18172,7 +18172,9 @@ with pkgs;
 
   libantlr3c = callPackage ../development/libraries/libantlr3c {};
 
-  libaom = callPackage ../development/libraries/libaom { };
+  libaom = callPackage ../development/libraries/libaom {
+    libjxl = libjxl.override { enableDocs = false; }; # libavif -> libaom -> libjxl -> graphviz -> gd -> libavif
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18176]:
```suggestion
    # escape dependency chain with libavif
    libjxl = libjxl.override { enableDocs = false; };
```
Lets not hardcode the path as that might change in the future

**Comment by SuperSandro2000** [comment]
otherwise LGTM

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -18172,7 +18172,10 @@ with pkgs;
 
   libantlr3c = callPackage ../development/libraries/libantlr3c {};
 
-  libaom = callPackage ../development/libraries/libaom { };
+  libaom = callPackage ../development/libraries/libaom {
+    # Remove circular dependency for libavif
+    libjxl = libjxl.override { enableDocs = false; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18177]: That flag is called buildDocs.

Comment (dali99) [reply pkgs/top-level/all-packages.nix:18177]: that's embarrassing, I thought I tested evaling it

Comment (dali99) [reply pkgs/top-level/all-packages.nix:18177]: I guess lazy loading means it never actually checked once it started building?

anyways, fixed

**Comment by trofi** [comment]
Bisect says 397724176b474080a09422fdc29d3fc85cd43f2a `libaom: Add support for butteraugli and vmaf tunes` broke `gst_all_1.gst-plugins-bad`:

```
$ nix build -f. gst_all_1.gst-plugins-bad -L
...
gst-plugins-bad> Run-time dependency openssl found: YES 1.1.1o
gst-plugins-bad> Run-time dependency libass found: YES 0.15.2
gst-plugins-bad> Run-time dependency aom found: NO
gst-plugins-bad> ext/aom/meson.build:1:0: ERROR: Could not generate cargs for aom:
gst-plugins-bad> Package libvmaf was not found in the pkg-config search path.
gst-plugins-bad> Perhaps you should add the directory containing `libvmaf.pc'
gst-plugins-bad> to the PKG_CONFIG_PATH environment variable
gst-plugins-bad> Package 'libvmaf', required by 'aom', not found
gst-plugins-bad> A full log can be found at /build/gst-plugins-bad-1.20.1/build/meson-logs/meson-log.txt
error: builder for '/nix/store/73pg3a1wmznsk0ypwny1jz4ympj3a16z-gst-plugins-bad-1.20.1.drv' failed with exit code 1;
       last 10 log lines:
       > Run-time dependency aom found: NO
       >
       > ext/aom/meson.build:1:0: ERROR: Could not generate cargs for aom:
       > Package libvmaf was not found in the pkg-config search path.
       > Perhaps you should add the directory containing `libvmaf.pc'
       > to the PKG_CONFIG_PATH environment variable
       > Package 'libvmaf', required by 'aom', not found
       >
       >
       > A full log can be found at /build/gst-plugins-bad-1.20.1/build/meson-logs/meson-log.txt
       For full logs, run 'nix log /nix/store/73pg3a1wmznsk0ypwny1jz4ympj3a16z-gst-plugins-bad-1.20.1.drv'.
```

**Comment by trofi** [comment]
I suspect it's because `libaom.dev` is missing `libvmaf.dev` reference in it's pkgconfig files.

**Comment by dali99** [comment]
@trofi I added libvmaf to `propagatedBuildInputs` and that makes gst-plugins-bad build again. Though I kind of doubt gstreamer actually uses the vmaf support, so a additionally we could maybe disable vmaf for it

**Comment by trofi** [comment]
Adding `propagatedBuildInputs` sounds good.

I think having multiple instances of a library (libaom{enableVmaf=true;} and libaom{enableVmaf=false;}) in a single closure of a library is not very safe. Who know which of two will be loaded first in RAM. I'd suggest just making `libaom{enableVmaf=true;}` case to work and keep gst-plugins-bad just use it.

**Pull request merged**